puppet/zulip/manifests/nginx.pp

```diff
@@ -29,23 +29,32 @@
     notify  => Service['nginx'],
   }
 
+  # Configuration for how uploaded files are served.  The default is
+  # to serve uploaded files using using the `nginx` `internal`
+  # feature via django-sendfile2, which basically does an internal
+  # redirect and returns the file content from nginx in an
+  # HttpResponse that would otherwise have been a redirect.
+  #
+  # For installations using S3 to serve uploaded files, we want Django
+  # to handle the /serve_uploads route, so that it can do authentication
+  # and serve a redirect.
   $no_serve_uploads = zulipconf('application_server', 'no_serve_uploads', '')
-  if $no_serve_uploads != '' {
-    # If we're not serving uploads locally, set the appropriate API headers for it.
-    $uploads_route = 'puppet:///modules/zulip/nginx/zulip-include-maybe/uploads-route.noserve'
+  if $no_serve_uploads == '' {
+    file { '/etc/nginx/zulip-include/app.d/uploads-internal.conf':
+      ensure  => file,
+      require => Package[$zulip::common::nginx],
+      owner   => 'root',
+      group   => 'root',
+      mode    => '0644',
+      notify  => Service['nginx'],
+      source  => 'puppet:///modules/zulip/nginx/zulip-include-maybe/uploads-internal.conf',
+    }
   } else {
-    $uploads_route = 'puppet:///modules/zulip/nginx/zulip-include-maybe/uploads-route.internal'
+    file { '/etc/nginx/zulip-include/app.d/uploads-internal.conf':
+      ensure  => absent,
+    }
   }
```

Comment:
We probably also need an `ensure => absent` for the old `/etc/nginx/zulip-include/uploads.route`, right?

Reply:
Not strictly necessary, since nothing includes it anymore, but good for cleanup, yes.